numpy/lib/shape_base.py

```diff
@@ -133,10 +133,12 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     if not isinstance(res, matrix):
         buff = res.__array_prepare__(buff)
 
-    # save the first result, then compute and save all remaining results
-    buff[ind0] = res
+    # save the first result, then compute and save all remaining results.
+    # 0d arrays need to be converted to scalars for this to work for object
+    # arrays (gh-8642), which can be done with `[()]`
+    buff[ind0] = res[()]
     for ind in inds:
-        buff[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))
+        buff[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))[()]
```

Comment:
This could also have been fixed by changing this line to `buff[ind + np.s_[...]]`, and leaving the rest untouched. Probably better for not going through a scalar, and avoiding 7267?